gnucash/gnome/gnc-budget-view.c

```diff
@@ -913,6 +914,15 @@ get_negative_color (void)
     return gdk_rgba_to_string(&color);
 }
 
+static void
+get_flag_color (GdkRGBA *color)
+{
+    GtkWidget *label = gtk_label_new ("Color");
+    GtkStyleContext *context = gtk_widget_get_style_context (GTK_WIDGET(label));
+    gtk_style_context_add_class (context, "flag-color");
+    gtk_style_context_get_color (context, GTK_STATE_FLAG_NORMAL, color);
+    gtk_widget_destroy(label);
+}
```

Comment:
While this works for the most basic scenario determining the flag_color from an arbitrary and temporary widget the user never interacts with is fragile.
You should query the style context of the exact widget that will be using these style properties to render in order to allow the user more fine-grained control over styling.

The way it's implemented now only it's not possible for a user to style as follows:
```
even-row > .flag-color {
    color: pink;
}

odd-row > .flag-color {
    color: magenta;
}
```

Currently this is a rather hypothetical example, it's just to show the point. If any css selector is used other than just the .flag-color class, the css snippet will not be found in your test.

Secondly `flag-color` is not a good name for a css class selector. It already presumes only color will be set and prevents future expansion. Another user may want to add properties (and accompanying code) to add margins for example. flag-color would not really cover the class any more.

A better name could be 'notes-flag' or even 'bugdet-notes-indicator'.

Reply:
Thinking some more about this, it may even be better to query this color insed the gnc_cell_renderer_text_flag code as part of the initialization function. In that case the name should probably also become more generic, something like 'txt-renderer-flag' or something like that.

Doing it in the initialization function of the gnc_cell_renderer_text_flag widget allows a consumer of that widget to still override the color before rendering by setting the `flag-color-rgba` property of the widget. I personally would not recommend anyone to do so programmatically, but it keeps the renderer generic and customizable, with a css based default style.

Reply:
@gjanssens the flag renderer uses an attribute to define the color, using the same approach as the standard widget.
The renderer itself is not a widget, and thus it doesn’t have a widget context of it own. I don’t see how this can be done with a generic style.

Regarding setting the color in the initialization of the renderer, as there is no widget, you have no context and so no possibility to use css.

I agree that the “flag-color” approach has it limitations, but they are the same as in other places of gnucash (like getting the negative color).
Additionally, the tree widget doesn’t allow to do much customizing via css (the odd-row example you used is not supported), most customization must be done via attributes.

I don’t find a more generic way to do it. The only widget present is the whole treeview. Using that widget as context for the css wouldn’t work as it would set the “color” css attribute to the whole view, not just the notes label.

Reply:
It's a bit more complicated than I first thought indeed as the renderer doesn't have a widget.
And unfortunately Gtk doesn't expose its private GtkNode interface. That would have allowed you to add another css node to the budget GtkTreeView.

I still think we can come close to what I have in mind by adding the label widget directoy as a hidden child to the treeview. That way you'd have an extra css node users can style under the treeview instead of requiring users to style some pre-defined class. Have you investigated this possibility ?

That aside, I also would like you to change the class name to something that's more generic than color. There may be other useful css settings in the future and changing names after they have been introduced is a pain.

Reply:
I don’t think I follow your idea:

```
I still think we can come close to what I have in mind by adding the label widget directoy as a hidden child to the treeview.
```

There is no “label widget”, just the renderer, which does not create any child widget. It just draws on the treeview canvas.
Creating child widgets defeats the whole “renderer” concept of the treeview.
 I’ll change the class name and rebase, but I don’t find a different approach.

Reply:
Rebased and changed css class name from `flag-color` to `bugdet-notes-indicator`.

Reply:
Regarding your comment:

```
That way you'd have an extra css node users can style under the treeview instead of requiring users to style some pre-defined class. 
```
I'm not sure sure I understand. You mean that instead of styling a *predefined* class, the should style a non-standard hierarchy like:
```
treeview label { 
   color: ....
```

I think it *might* be possible, but I really don't see how that is better or more intuitive.
In fact I find it is much more difficult to understand what it targets.

In general I see that in GnuCash all styling is done with *predefined classes*:
```
*.register-foreground {
*.register-header {
.negative-numbers {
```
why use such a different approach in this case only?

Reply:
Ok, I took a pause to give this some more thought.

First an answer to your question "why use a different approach in this case only?"
Because  the case is different. The examples you mention define these style classes inside the widgets to which they refer, where your style class is defined outside of that widget. That is the core of my issue with your approach.

However stepping back I think we may be over-engineering this whole thing. I actually think that creating a custom renderer only to be able to add a visual clue to a note may be a bit too much. A treeview column can have more than one renderer. So why not just add two renderers to each cell ? One to display the number and the other to display a small indicator (which can be a small icon) whenever a note is attached to the value. It seems to me this would be more straightforward and reusing already functionality in gtk.

If you like the icon to follow the current theme you can pick one of [the standard icons](https://standards.freedesktop.org/icon-naming-spec/icon-naming-spec-latest.htmlurl). You could also find the names by opening the glade ui editor and create an GtkImage widget. Then the Image selector will list all stock images you can use directly in your image renderer.


Aside from that, I have built your branch and played with it. It looks pretty nice. I would only suggest to add the "Note" feature to the "Edit" menu, not only the tool bar.

Reply:
Hi @gjanssens,
I don’t think that adding an icon to the cell is a good idea to just signal a note.
It would be too intrusive, potentially cluttering the grid if there are many notes, and it would take too much space, making columns wider and so less columns would fit on the screen.
I tried not to reinvent the wheel here, and used the most “standard” and proven approach used by Excel (and so I think also more intuitive to users) of using just a small triangle to mark notes.
As the use case is very specific, I don’t think the icon should be customizable (apart from color).

Regarding the custom renderer approach, the truth is that I’m not a gtk expert and I just couldn’t find a way to use two std renderers (text renderer + pixbuf renderer) and make the pixbuf draw the triangle exactly at the corner and without taking any extra space in the cell. 
If you can give me some directions, I could give it a new try.

Anyhow the custom renderer is rather simple and doesn’t add too much complexity nor size.

Reply:
> I tried not to reinvent the wheel here, and used the most “standard” and proven approach used by Excel (and so I think also more intuitive to users) of using just a small triangle to mark notes.

That assumes users are using Excel. I don't and I think most linux user don't. I haven't seen this in LibreOffice yet.
I mostly mention this to point out we should be careful with our assumptions. Other than that I think a triangle is fine. I didn't suggest the "icon" had to be a full fledged icon. An image of a triangle may have worked as well. However I have no idea how far you can configure two renderers to share the same rendering area. I suspect it's possible, but I have never experimented with this. Drawing an image in the top-right corner by itself is easy. One can simply set xalign to 1 (right aligned) and yalign to 0 (top aligned).

> Anyhow the custom renderer is rather simple and doesn’t add too much complexity nor size.

That is a matter of perspective of course... Even simple code can add bugs, and your code actually adds a potential memory leak already (I'll mark it in a minute). In addition it adds another code branch in the gnucash tree view (initialization with or without a custom renderer). In testing that would mean a factor two more of parameter combinations to check.

But none of that is really the point of my issue with this PR. My main issue is that the CSS selector you have set up for the flag color is not logically tied to the widget using it and the suggestion of using a combination of two renderers was one thought experiment to get rid of this. To be honest I don't believe either it will be such a clean solution.

Your custom cell renderer in itself is conceptually ok. We need to figure out a cleaner way to define the color(s) it should use. I am a strong proponent of reusing what's already available if it comes to styling. So here's another avenue to explore:
* The treeview uses different background colors for a cell (or row) that's selected and one that's not
* These colors come from the current css theme and hence should be carefully chosen by the theme designer
* So as colors for the flag triangle I propose to use the opposite colors: when a cell is not selected use the treeview's selection background color as flag color. For a selected cell use the treeview's normal background color as flag color. That way
1.  we stay in the theme
2. we should have flag triangles that are visible in all situations
3. you don't need the whole clutch to create our own theme overrides.

You can probably fine-tune it a little more even by using selected+disabled's  background color if the cell is disabled. This is typically slightly dimmed.

Reply:
Switched from css to using inverted "selected"/"normal" cell's background color for note's flag